internal/provider/resource_ably_key.go

```diff
@@ -0,0 +1,276 @@
+			},
+			"modified": {
+				Type:        types.Int64Type,
+				Computed:    true,
+				Description: "Unix timestamp representing the date and time of the last modification of the key.",
+			},
+		},
+	}, nil
+}
+
+// New resource instance
+func (r resourceKeyType) NewResource(_ context.Context, p tfsdk_provider.Provider) (tfsdk_resource.Resource, diag.Diagnostics) {
+	return resourceKey{
+		p: *(p.(*provider)),
+	}, nil
+}
+
+type resourceKey struct {
+	p provider
+}
+
+// Create a new resource
+func (r resourceKey) Create(ctx context.Context, req tfsdk_resource.CreateRequest, resp *tfsdk_resource.CreateResponse) {
+	// Checks whether the provider and API Client are configured. If they are not, the provider responds with an error.
+	if !r.p.configured {
+		resp.Diagnostics.AddError(
+			"Provider not configured",
+			"The provider hasn't been configured before apply",
+		)
+		return
+	}
+
+	// Gets plan values
+	var plan AblyKey
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	new_key := ably_control_go.NewKey{
+		Name:       plan.Name.Value,
+		Capability: plan.Capability,
+	}
+
+	// Creates a new Ably Key by invoking the CreateKey function from the Client Library
+	ably_key, err := r.p.client.CreateKey(plan.AppID.Value, &new_key)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating Resource",
+			"Could not create resource, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// Maps response body to resource schema attributes.
+	resp_key := AblyKey{
+		ID:         types.String{Value: ably_key.ID},
+		AppID:      types.String{Value: ably_key.AppID},
+		Name:       types.String{Value: ably_key.Name},
+		Key:        types.String{Value: ably_key.Key},
+		Capability: ably_key.Capability,
+		Status:     types.Int64{Value: int64(ably_key.Status)},
+		Created:    types.Int64{Value: int64(ably_key.Created)},
+		Modified:   types.Int64{Value: int64(ably_key.Modified)},
+	}
+
+	// Sets state for the new Ably App.
+	diags = resp.State.Set(ctx, resp_key)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Read resource
+func (r resourceKey) Read(ctx context.Context, req tfsdk_resource.ReadRequest, resp *tfsdk_resource.ReadResponse) {
+	// Gets the current state. If it is unable to, the provider responds with an error.
+	var state AblyKey
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Gets the Ably App ID and Ably API Key ID value for the resource
+	app_id := state.AppID.Value
+	key_id := state.ID.Value
+
+	// Fetches all Ably Keys for the Ably App. The function invokes the Client Library Keys() method.
+	keys, _ := r.p.client.Keys(app_id)
+
+	// Loops through apps and if account id and key id match, sets state.
+	for _, v := range keys {
+		if v.AppID == app_id && v.ID == key_id {
```

Comment:
Should we check for non revoked also?

Reply:
Or rather match anyway but if it's revoked error out?

Reply:
Added check for revoked.